Add basic support for Redshift and RDS Data APIs

pwithams opened the pull request:
Issue #798

## Description of changes
Added basic Redshift Data API Service and RDS Data Service by adding a new module `data_api`. This module contains:
1. `connector.py`: a base class that defines a basic data api interface along with some shared logic
2. `rds.py`: contains the RdsDataApi class as well as the `connect` and `read_sql_query` functions
3. `redshift.py`: contains the RedshiftDataApi class as well as the `connect` and `read_sql_query` functions

The new functionality can be used as follows:

```python
# Redshift
 con = wr.data_api.redshift.connect("clusterid123", "my_database", secret_arn="arn:aws:secretsmanager...")
 dataframe = wr.data_api.redshift.read_sql_query("SELECT * FROM public.my_table limit 10", con=con)

# RDS
 con = wr.data_api.rds.connect("arn:aws:rds...", "my_database", secret_arn="arn:aws:secretsmanager...")
 dataframe = wr.data_api.rds.read_sql_query("SELECT * FROM my_table limit 10", con=con)
```

Local unit tests were added at the end of `tests/test_moto.py` and integration tests using the CDK stack under `test_infra/` were added to `tests/test_data_api.py`.

To support RDS Data API, an Aurora serverless cluster was added to `test_infra/stacks/databases_stacks.py`.


## Discussion points
 - Currently only basic SQL string statements are supported and have been tested
 - Redshift Data API is asynchronous so the new module features a custom waiter that waits until a result is ready
 - RDS class is already synchronous but wait logic also exists in the case where a serverless cluster is in a paused state
 - Unit tests feature basic custom mocking of API responses
 - Integration tests only cover a few basic query examples

There is most likely a lot more work to be done but I think I have enough to get some feedback on the approach etc. For context I have experience with the Redshift Data API but not the RDS Data API. Support for that is more preemptive rather than reactive on my part and so may be missing a few things.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

jaidisido commented:
### AWS CodeBuild CI Report

* CodeBuild project: GitHubCodeBuild8756EF16-pDO66x4b9gEu
* Commit ID: e9c163ef5ba2f8705113bb58c81b5d6804b1b404
* Result: FAILED
* [Build Logs](https://8z22lm66x0.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=98243c84-993d-42b5-9c1f-226ab4609f84%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

jaidisido commented:
### AWS CodeBuild CI Report

* CodeBuild project: GitHubCodeBuild8756EF16-pDO66x4b9gEu
* Commit ID: a86bdce8e2b348c06500a7bcee854b7c19fe63f1
* Result: FAILED
* [Build Logs](https://8z22lm66x0.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=72124fd3-7c35-4fc8-96b8-6083da703033%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

pwithams commented:
I'm not exactly sure how the Code Build workflow is setup but it looks like the build failed due to:

```
>       parameters["redshift"]["secret_arn"] = cloudformation_outputs["RedshiftSecretArn"]
E       KeyError: 'RedshiftSecretArn'
```

The "RedshiftSecretArn" key is pulled from `test_infra/stacks/databases_stack.py`, which I updated as part of my changes:

```
$ git grep RedshiftSecretArn

test_infra/stacks/databases_stack.py:        cdk.CfnOutput(self, "RedshiftSecretArn", value=secret.secret_arn)
tests/conftest.py:    parameters["redshift"]["secret_arn"] = cloudformation_outputs["RedshiftSecretArn"]
```

I also added a serverless Aurora cluster via `setup_mysql_serverless()` to `test_infra/stacks/databases_stack.py` that is required for the RDS tests in `tests/test_data_api.py`.

However, my guess is that the test stack is not automatically deployed on each build (which makes sense) and so even though I updated the CDK code, those values/resources won't be available until the test stack is redeployed.

Let me know if I'm missing something or if there's anything else I can do in the meantime. Thanks.

kukushking commented:
Thanks for the contribution! Yes, your understanding is correct - test infrastructure is pre-set up in the account and does not automatically update on PR's for security reasons. I/team will look into your PR and update the infra if there are no major concerns with it. Have you tried testing in your own AWS acccount?

pwithams commented:
Thanks for the response - yes I've run the `tests/test_data_api.py` tests in my account after running `./deploy_base.sh` and running `./deploy_databases.sh` with the CDK changes (added Aurora serverless cluster for RDS data Api test and the Redshift secret arn as output for the Redshift data Api test).

Obviously it costs me money to run in my own account so I'd favour CI/CD tests where possible. However, I don't mind running in my own account every now again.

jaidisido commented:
### AWS CodeBuild CI Report

* CodeBuild project: GitHubCodeBuild8756EF16-pDO66x4b9gEu
* Commit ID: a86bdce8e2b348c06500a7bcee854b7c19fe63f1
* Result: FAILED
* [Build Logs](https://8z22lm66x0.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=f875ade4-f222-4d92-88e5-3959305ff8a9%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

kukushking commented on awswrangler/data_api/rds.py:
```diff
@@ -0,0 +1,151 @@
+"""RDS Data API Connector."""
+from __future__ import annotations
```
This appears to be available [only starting from 3.7](https://www.python.org/dev/peps/pep-0563/#enabling-the-future-behavior-in-python-3-7). 

Unit/integration tests currently run against 3.7 only so wouldn't catch that but we still try to keep 3.6 supported and test every release against 3.6, 3.7, 3.8 & 3.9. Can we remove it and replace string type hints where applicable?

pwithams replied:
Good catch, yes I'll fix that. Looks like the latest test failures are due to the code build job needing new IAM permissions for the Data API calls (i.e. ExecuteStatement, etc.)

pwithams replied:
Updated now. I was able to remove the dependency by just moving the functions to the end of the files below the class definition.

As those functions are the main interface I thought it would be helpful for them to be at the top of the file rather than buried at the end, but it was just a "nice to have". Could probably keep them at the top and use string hints, but moving them to the end of the file seemed like the simpler/cleaner solution.

kukushking replied:
Thanks! Yes, I agree it's cleaner. I added `"rds-data:ExecuteStatement" & "redshift-data:ExecuteStatement"` & re-triggered the tests - let's see if there is anything else it needs.

jaidisido commented:
### AWS CodeBuild CI Report

* CodeBuild project: GitHubCodeBuild8756EF16-pDO66x4b9gEu
* Commit ID: a86bdce8e2b348c06500a7bcee854b7c19fe63f1
* Result: FAILED
* [Build Logs](https://8z22lm66x0.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=10e50992-26c4-45cb-9411-7360d2169569%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

jaidisido commented:
### AWS CodeBuild CI Report

* CodeBuild project: GitHubCodeBuild8756EF16-pDO66x4b9gEu
* Commit ID: a86bdce8e2b348c06500a7bcee854b7c19fe63f1
* Result: FAILED
* [Build Logs](https://8z22lm66x0.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=688b5a46-245c-4ba1-8959-0b7d60357d83%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

jaidisido commented:
### AWS CodeBuild CI Report

* CodeBuild project: GitHubCodeBuild8756EF16-pDO66x4b9gEu
* Commit ID: 8bb568e9cd10dcd8d829a57a3311b1710e24b7e5
* Result: FAILED
* [Build Logs](https://8z22lm66x0.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=b2da42a1-aef2-421a-beb2-1370108a97c4%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

jaidisido commented:
### AWS CodeBuild CI Report

* CodeBuild project: GitHubCodeBuild8756EF16-pDO66x4b9gEu
* Commit ID: 8bb568e9cd10dcd8d829a57a3311b1710e24b7e5
* Result: FAILED
* [Build Logs](https://8z22lm66x0.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=c48d83b1-3b59-43b5-be40-87dd93aaeba5%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

jaidisido commented:
### AWS CodeBuild CI Report

* CodeBuild project: GitHubCodeBuild8756EF16-pDO66x4b9gEu
* Commit ID: 8bb568e9cd10dcd8d829a57a3311b1710e24b7e5
* Result: FAILED
* [Build Logs](https://8z22lm66x0.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=4d8d1250-8d44-45c0-8196-89b4e569a31f%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

jaidisido commented:
### AWS CodeBuild CI Report

* CodeBuild project: GitHubCodeBuild8756EF16-pDO66x4b9gEu
* Commit ID: 8bb568e9cd10dcd8d829a57a3311b1710e24b7e5
* Result: SUCCEEDED
* [Build Logs](https://8z22lm66x0.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=76f6e07e-6fa8-4a92-aec0-57d386e7f123%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

jaidisido commented:
### AWS CodeBuild CI Report

* CodeBuild project: GitHubCodeBuild8756EF16-pDO66x4b9gEu
* Commit ID: 4ef1e274a659dbc1057fa58de8ce5d496ea020d5
* Result: SUCCEEDED
* [Build Logs](https://8z22lm66x0.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=01362dcc-befa-44e8-8ac3-e1379de3809c%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->
